File: AntikytheraOpenGLPrototype_v3/Classes/FlyThroughCamera.m
//
//  FlyThroughCamera.m
//  AntikytheraOpenGLPrototype
//
//  Created by Matt Ricketson on 4/8/10.
//  Copyright 2010 Apple Inc. All rights reserved.
//

#import "FlyThroughCamera.h"


@implementation FlyThroughCamera

- (void) updateViewpoint {
	static GLfloat rot = 0.0;
	
	glLoadIdentity();
	
	glTranslatef(0.0f,0.0f,-100.0+50*sin(-rot/100));
	glRotatef(-rot, 1.0, 1.0, 1.0);
	
	glClearColor(0.0, 0.0, 0.0, 1.0);
	glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
	
	static NSTimeInterval lastDrawTime;
    if (lastDrawTime)
    {
        NSTimeInterval timeSinceLastDraw = [NSDate timeIntervalSinceReferenceDate] - lastDrawTime;
		rot+=10 * timeSinceLastDraw;
		//		NSLog(@"%f",timeSinceLastDraw);
	}
    lastDrawTime = [NSDate timeIntervalSinceReferenceDate];
}

@end

File: AntikytheraOpenGLPrototype_v3/Classes/FlyThroughCamera.swift
//
//  FlyThroughCamera.swift
//  AntikytheraOpenGLPrototype
//
//  Created by Matt Ricketson on 4/8/10.
//  Copyright 2010 Apple Inc. All rights reserved.
//

import Metal
import MetalKit

class FlyThroughCamera: NSObject, CameraViewpoint {

    var rot: Float = 0.0
    var lastDrawTime: TimeInterval = 0

    func updateViewpoint() {
        guard let renderer = MetalRenderContext.shared.renderer else { return }

        renderer.loadIdentity()
        renderer.translate(x: 0.0, y: 0.0, z: -100.0 + 50 * sin(-rot / 100))
        renderer.rotate(angle: -rot, x: 1.0, y: 1.0, z: 1.0)

        if lastDrawTime != 0 {
            let timeSinceLastDraw = Date.timeIntervalSinceReferenceDate - lastDrawTime
            rot += Float(10 * timeSinceLastDraw)
        }
        lastDrawTime = Date.timeIntervalSinceReferenceDate
    }

}
